javascript/session_test.ts

```diff
@@ -0,0 +1,137 @@
+import {
+    assertEquals,
+} from "https://deno.land/std/testing/asserts.ts";
+import {
+    Buffer,
+} from "https://deno.land/std/io/mod.ts";
+
+// @ts-ignore
+import * as channel from "./channel.ts";
+import * as session from "./session.ts";
+import * as api from "./api.ts";
+import * as util from "./util.ts";
+
+class Conn implements api.IConn {
+    conn: Deno.Conn;
+
+    constructor(conn: Deno.Conn) {
+        this.conn = conn;
+    }
+
+    async read(len: number): Promise<Uint8Array | undefined> {
+        let buff = new Uint8Array(len);
+        // TODO does this need to loop if we read less than the expected amount?
+        let n = await this.conn.read(buff);
+        if (n == null) {
+            return undefined;
+        }
+        if (buff.byteLength > n) {
+            buff = buff.slice(0, n);
+        }
+        return buff;
+    }
+
+    write(buffer: Uint8Array): Promise<number> {
+        return this.conn.write(buffer)
+    }
+
+    close(): Promise<void> {
+        // XXX this errors with:
+        // BadResource: Bad resource ID if the connection was already closed
+        // because the other end disconnected. Should we catch it here, or
+        // let that error bubble up?
+        this.conn.close();
```

Comment:
Any thoughts on whether to propagate this error, or suppress it here?

Reply:
I'm always annoyed with already closed errors.

Reply:
Plus I know there was some uncatchable exceptions that bothered me even more so ... but I say suppress it if it is Bad Resource ID.